File: authorization/src/main/java/com/example/alexeyglushkov/authorization/Auth/Authorizer.java
package com.example.alexeyglushkov.authorization.Auth;

import androidx.annotation.NonNull;
import io.reactivex.Observable;
import io.reactivex.Single;

/**
 * Created by alexeyglushkov on 31.10.15.
 */
public interface Authorizer {
    Single<AuthCredentials> authorize();
    void signCommand(ServiceCommand command, AuthCredentials credentials);

    void setServiceCommandProvider(ServiceCommandProvider provider);
    void setServiceCommandRunner(ServiceCommandRunner runner);

    class AuthError extends Error {
        private static final long serialVersionUID = 6206983256074915330L;

        public enum Reason {
            InnerError,
            UnknownError,
            Cancelled,
            NotAuthorized // for services
        }

        protected @NonNull Reason reason;

        public @NonNull Reason getReason() {
            return reason;
        }

        public AuthError(@NonNull Reason reason, Throwable throwable) {
            super(throwable);
            this.reason = reason;
        }

        public AuthError(String detailMessage, @NonNull Reason reason, Throwable throwable) {
            super(detailMessage, throwable);
            this.reason = reason;
        }
    }
}

File: authorization/src/main/java/com/example/alexeyglushkov/authorization/Auth/Authorizer.kt
package com.example.alexeyglushkov.authorization.Auth

import io.reactivex.Single
import java.lang.Exception

/**
 * Created by alexeyglushkov on 31.10.15.
 */
interface Authorizer {
    suspend fun authorize(): AuthCredentials
    fun signCommand(command: ServiceCommand<*>, credentials: AuthCredentials)
    fun setServiceCommandProvider(provider: ServiceCommandProvider)
    fun setServiceCommandRunner(runner: ServiceCommandRunner)

    class AuthError : Exception {
        enum class Reason {
            InnerError,
            UnknownError,
            Cancelled,
            NotAuthorized // for services
        }

        var reason: Reason
            protected set

        constructor(reason: Reason, throwable: Throwable?) : super(throwable) {
            this.reason = reason
        }

        constructor(detailMessage: String?, reason: Reason, throwable: Throwable?) : super(detailMessage, throwable) {
            this.reason = reason
        }

        companion object {
            private const val serialVersionUID = 6206983256074915330L
        }
    }
}